src/Core/Framework/App/Url/AppUrlVerifier.php

```diff
@@ -0,0 +1,279 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Framework\App\Url;
+
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\Clock\ClockInterface;
+use Shopware\Core\Framework\App\AppException;
+use Shopware\Core\Framework\App\ShopId\Fingerprint\AppUrl;
+use Shopware\Core\Framework\App\ShopId\ShopId;
+use Shopware\Core\Framework\Log\Package;
+use Symfony\Component\Clock\NativeClock;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Lock\Exception\LockAcquiringException;
+use Symfony\Component\Lock\Exception\LockConflictedException;
+use Symfony\Component\Lock\LockFactory;
+use Symfony\Component\Lock\LockInterface;
+use Symfony\Contracts\Cache\CacheInterface;
+use Symfony\Contracts\Cache\ItemInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @internal
+ */
+#[Package('framework')]
+class AppUrlVerifier
+{
+    public const VERIFICATION_RESULT_CACHE_KEY = 'app_url_verification_result';
+    private const VERIFICATION_CACHE_KEY_PREFIX = 'app_url_verify-';
+    private const BACK_OFF = [
+        1 => 60 * 5,  // 5 minutes
+        2 => 60 * 15, // 15 minutes
+    ];
+    private const MAX_TRIES = 3;
+    private const VERIFY_PATH = '/api/app-system/shop/verify';
+
+    public function __construct(
+        private readonly string $appEnv,
+        private readonly string $shopwareVersion,
+        private readonly CacheItemPoolInterface&CacheInterface $cache,
+        private readonly HttpClientInterface $httpClient,
+        private readonly LockFactory $lockFactory, /** Should be a lock with ttl support - otherwise locks might not expire and verification will not be performed */
```

Comment:
is that true? I don't see where we rely on expiry of the lock, we'll release it in finally block, do you mean the cache? because there we use the expiry to perfrom the verifacation every 24h, doesn't every cache implementation in symfony support expiry?

Reply:
It's more about if the process crashes and doesn't get released by us - and no not all locks support expiring: https://symfony.com/doc/current/components/lock.html#available-stores